codegen/api/src/main/java/io/sundr/codegen/api/CodeGenerator.java

```diff
@@ -0,0 +1,160 @@
+    private final Consumer<T> onSkip;
+
+    private Builder(Class<T> type) {
+      this(type, null, null, null, null, null);
+    }
+
+    private Builder(Class<T> type, Function<T, Writer> output, Function<T, String> identifier, Function<T, String> renderer,
+        Predicate<T> skip,
+        Consumer<T> onSkip) {
+      this.type = type;
+      this.output = output;
+      this.identifier = identifier;
+      this.renderer = renderer;
+      this.skip = skip;
+      this.onSkip = onSkip;
+    }
+
+    public Builder<T> withOutput(Output<T> output) {
+      return withOutput(output.getFunction());
+    }
+
+    public Builder<T> withOutput(Function<T, Writer> output) {
+      return new Builder<>(type, output, identifier, renderer, skip, onSkip);
+    }
+
+    public Builder<T> withIdentifier(Identifier<T> identifier) {
+      return withIdentifier(identifier.getFunction());
+    }
+
+    public Builder<T> withIdentifier(Function<T, String> identifier) {
+      return new Builder<>(type, output, identifier, renderer, skip, onSkip);
+    }
+
+    public Builder<T> withRenderer(Renderer<T> renderer) {
+      return withRenderer(renderer.getFunction());
+    }
+
+    public Builder<T> withRenderer(Function<T, String> renderer) {
+      return new Builder<>(type, output, identifier, renderer, skip, onSkip);
+    }
+
+    public Builder<T> skipping(Predicate<T> skip) {
+      return new Builder<>(type, output, identifier, renderer, skip, onSkip);
+    }
+
+    public CodeGenerator<T> build() {
+      return new CodeGenerator<>(type, output, identifier, renderer, skip, onSkip);
+    }
+
+    public boolean generate(T... items) {
+      return build().generate(items);
+    }
+  }
+
+  private CodeGenerator(Class<T> type, Function<T, Writer> output, Function<T, String> identifier, Function<T, String> renderer,
+      Predicate<T> skip, Consumer<T> onSkip) {
+    this.type = type;
+    this.output = output != null ? output : new SystemOutput<T>().getFunction();
+    this.identifier = identifier != null ? identifier
+        : Identifiers.findIdentifier(type).map(Identifier::getFunction).orElse(o -> String.valueOf(o.hashCode()));
+    this.renderer = renderer != null ? renderer : Renderers.findRenderer(type).map(Renderer::getFunction).orElse(null);
+    this.skip = skip != null ? skip : Predicates.distinct(t -> this.identifier.apply(t));
+    this.onSkip = onSkip != null ? onSkip : ignore;
+  }
+
+  public static <T> Builder<T> newGenerator(Class<T> type) {
+    return new Builder<T>(type);
+  }
+
+  /**
+   * Generate one or more items
+   * 
+   * @param <T> the type of items to generate
+   * @param items the items to generate
+   * @return true if generation was succesful, false otherwise
+   */
+  public boolean generate(T... items) {
+    if (renderer == null) {
+      throw new IllegalStateException("Renderer should not be null.");
+    }
+
+    if (output == null) {
+      throw new IllegalStateException("Output should not be null.");
+    }
```

Comment:
Not sure if this is stupid, but if these variables are final, why not throw the exception/perform validation in the constructor?